sopel/deprecated.py

```diff
@@ -0,0 +1,178 @@
+        func 2
+        >>> func3()
+        Deprecated since 7.0, will be removed in 8.0: obsolete
+        File "<stdin>", line 1, in <module>
+        func 3
+
+    The ``stack_frame`` argument can be used to choose which stack frame is
+    printed along with the message text. By default, this decorator prints the
+    most recent stack frame (the last entry in the list, ``-1``),
+    corresponding to where the decorated function itself was called. However,
+    in certain cases such as deprecating conditional behavior within an object
+    constructor, it can be useful to show a less recent stack frame instead.
+
+    .. note::
+
+        There is nothing that prevents this decorator to be used on a class's
+        method, or on any existing callable.
+
+    .. versionadded:: 7.0
+        Parameters ``reason``, ``version``, and ``removed_in``.
+
+    .. versionadded:: 7.1
+        The ``warning_in`` and ``stack_frame`` parameters.
+
+    """
+    if not any([reason, version, removed_in, warning_in, func]):
+        # common usage: @deprecated()
+        return deprecated
+
+    if callable(reason):
+        # common usage: @deprecated
+        return deprecated(func=reason)
+
+    if func is None:
+        # common usage: @deprecated(message, version, removed_in)
+        def decorator(func):
+            return deprecated(
+                reason, version, removed_in, warning_in, stack_frame, func)
+        return decorator
+
+    # now, we have everything we need to have:
+    # - message is not a callable (could be None)
+    # - func is not None
+    # - version and removed_in can be None but that's OK
+    # so now we can return the actual decorated function
+
+    message = reason or getattr(func, '__name__', '<anonymous-function>')
+
+    template = 'Deprecated: {message}'
+    if version and removed_in:
+        template = (
+            'Deprecated since {version}, '
+            'will be removed in {removed_in}: '
+            '{message}')
+    elif version:
+        template = 'Deprecated since {version}: {message}'
+    elif removed_in:
+        template = 'Deprecated, will be removed in {removed_in}: {message}'
+
+    text = template.format(
+        message=message, version=version, removed_in=removed_in)
+
+    @functools.wraps(func)
+    def deprecated_func(*args, **kwargs):
+        if not (warning_in and
+                parse_version(warning_in) >= parse_version(__version__)):
+            original_frame = inspect.stack()[-stack_frame]
+            mod = inspect.getmodule(original_frame[0])
+            module_name = None
+            if mod:
+                module_name = mod.__name__
+            if module_name:
+                if module_name.startswith('sopel.'):
+                    # core, or core plugin
+                    logger = logging.getLogger(module_name)
+                else:
+                    # probably a plugin; use Sopel's public API for getting the
+                    # logger for a plugin
```

Comment:
Not true any more. This just uses `logging` directly.

```suggestion
                    # probably a plugin; try to handle most cases sanely
```